src/Components/Web.JS/src/Platform/WebAssemblyResourceLoader.ts

```diff
@@ -0,0 +1,195 @@
+import { toAbsoluteUri } from '../Services/NavigationManager';
+
+export class WebAssemblyResourceLoader {
+  private usedCacheKeys: { [key: string]: boolean } = {};
+  private networkLoads: { [name: string]: LoadLogEntry } = {};
+  private cacheLoads: { [name: string]: LoadLogEntry } = {};
+
+  static async initAsync(): Promise<WebAssemblyResourceLoader> {
+    const bootConfigResponse = await fetch('_framework/blazor.boot.json', {
+      method: 'GET',
+      credentials: 'include'
+    });
+
+    const relativeBaseHref = document.baseURI.substring(document.location.origin.length);
+    const cacheName = `blazor-resources-${relativeBaseHref}`;
+    return new WebAssemblyResourceLoader(
+      await bootConfigResponse.json(),
+      await caches.open(cacheName));
+  }
+
+  constructor (public readonly bootConfig: BootJsonData, private cache: Cache)
+  {
+  }
+
+  loadResources(resources: ResourceList, url: (name: string) => string): LoadingResource[] {
+    return Object.keys(resources)
+      .map(name => this.loadResource(name, url(name), resources[name]));
+  }
+
+  loadResource(name: string, url: string, contentHash: string): LoadingResource {
+    const networkFetchOptions: RequestInit = { cache: 'no-cache' };
+    if (!this.bootConfig.cacheBootResources) {
+      // Bypass the entire cache flow, including checking hashes, etc.
+      // This gives developers an easy opt-out if they don't like anything about the default cache mechanism
+      const response = fetch(url, networkFetchOptions);
+      const data = response.then(r => r.clone().arrayBuffer());
```

Comment:
Added comment:

```
      // Cloning represents the response as two separate streams so we can process it twice independently
      // (once for hash checking, once for streaming compilation). Without this, there would be a read error.
```

Reply:
Yes, sorry. I should have been clearer. We are returning the response here, so we shouldn't need to clone it, as we are not going to cache it. Am I missing something obvious here?

Reply:
We're returning the response so it can be read externally, *and* a promise that resolves with its data. These two separate actions can't be done on a single response stream, hence the need to clone it.

However when I go back to this and make use of `fetch(..., { integrity: ... })`, it might be that quite a bit of this complexity goes away. There might be no need to be reading the response data separately. I'll find out when I try it :)

Reply:
It has got much simpler now I've changed to use `integrity`.

There is still one place where the response has to be cloned: before we put it into the cache. This is to be expected since we need to read the body (for caching) separately from returning it upstream. The "return from cache" code path, however, no longer has to clone.